// @applitools/eyes-selenium/dist/spec-driver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapDriver = exports.build = exports.performAction = exports.getElementText = exports.getElementAttribute = exports.getElementRegion = exports.getOrientation = exports.waitUntilDisplayed = exports.scrollIntoView = exports.type = exports.hover = exports.click = exports.takeScreenshot = exports.visit = exports.getUrl = exports.getTitle = exports.getDriverInfo = exports.setWindowSize = exports.getWindowSize = exports.findElements = exports.findElement = exports.childContext = exports.parentContext = exports.mainContext = exports.executeScript = exports.isEqualElements = exports.isStaleElementError = exports.transformSelector = exports.transformDriver = exports.isSelector = exports.isElement = exports.isDriver = void 0;
const utils = __importStar(require("@applitools/utils"));
const legacy = __importStar(require("./legacy"));
const byHash = ['className', 'css', 'id', 'js', 'linkText', 'name', 'partialLinkText', 'tagName', 'xpath'];
function extractElementId(element) {
    return element.getId();
}
function isDriver(driver) {
    return utils.types.instanceOf(driver, 'WebDriver');
}
exports.isDriver = isDriver;
function isElement(element) {
    return utils.types.instanceOf(element, 'WebElement');
}
exports.isElement = isElement;
function isSelector(selector) {
    if (!selector)
        return false;
    return utils.types.has(selector, ['using', 'value']) || byHash.includes(Object.keys(selector)[0]);
}
exports.isSelector = isSelector;
function transformDriver(driver) {
    driver.getExecutor().defineCommand('getSessionDetails', 'GET', '/session/:sessionId');
    driver.getExecutor().defineCommand('getOrientation', 'GET', '/session/:sessionId/orientation');
    driver.getExecutor().defineCommand('getSystemBars', 'GET', '/session/:sessionId/appium/device/system_bars');
    driver.getExecutor().defineCommand('performTouch', 'POST', '/session/:sessionId/touch/perform');
    if (process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3') {
        driver.getExecutor().defineCommand('switchToParentFrame', 'POST', '/session/:sessionId/frame/parent');
    }
    return driver;
}
exports.transformDriver = transformDriver;
function transformSelector(selector) {
    if (utils.types.isString(selector)) {
        return { css: selector };
    }
    else if (utils.types.has(selector, 'selector')) {
        if (!utils.types.isString(selector.selector))
            return selector.selector;
        if (!utils.types.has(selector, 'type'))
            return { css: selector.selector };
        if (selector.type === 'css')
            return { css: selector.selector };
        else
            return { using: selector.type, value: selector.selector };
    }
    return selector;
}
exports.transformSelector = transformSelector;
function isStaleElementError(error) {
    if (!error)
        return false;
    error = error.originalError || error;
    return error instanceof Error && error.name === 'StaleElementReferenceError';
}
exports.isStaleElementError = isStaleElementError;
async function isEqualElements(_driver, element1, element2) {
    if (!element1 || !element2)
        return false;
    const elementId1 = await extractElementId(element1);
    const elementId2 = await extractElementId(element2);
    return elementId1 === elementId2;
}
exports.isEqualElements = isEqualElements;
async function executeScript(driver, script, arg) {
    return driver.executeScript(script, arg);
}
exports.executeScript = executeScript;
async function mainContext(driver) {
    await driver.switchTo().defaultContent();
    return driver;
}
exports.mainContext = mainContext;
async function parentContext(driver) {
    if (process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3') {
        const { Command } = require('selenium-webdriver/lib/command');
        await driver.schedule(new Command('switchToParentFrame'));
        return driver;
    }
    await driver.switchTo().parentFrame();
    return driver;
}
exports.parentContext = parentContext;
async function childContext(driver, element) {
    await driver.switchTo().frame(element);
    return driver;
}
exports.childContext = childContext;
async function findElement(driver, selector, parent) {
    try {
        const root = parent !== null && parent !== void 0 ? parent : driver;
        return await root.findElement(selector);
    }
    catch (err) {
        if (err.name === 'NoSuchElementError')
            return null;
        else
            throw err;
    }
}
exports.findElement = findElement;
async function findElements(driver, selector, parent) {
    const root = parent !== null && parent !== void 0 ? parent : driver;
    return root.findElements(selector);
}
exports.findElements = findElements;
async function getWindowSize(driver) {
    try {
        const window = driver.manage().window();
        if (utils.types.isFunction(window.getSize)) {
            return await window.getSize();
        }
        else if (utils.types.isFunction(window.getRect)) {
            const rect = await window.getRect();
            return { width: rect.width, height: rect.height };
        }
    }
    catch (err) {
        const cmd = require('selenium-webdriver/lib/command');
        return driver.execute(new cmd.Command(cmd.Name.GET_WINDOW_SIZE).setParameter('windowHandle', 'current'));
    }
}
exports.getWindowSize = getWindowSize;
async function setWindowSize(driver, size) {
    const window = driver.manage().window();
    if (utils.types.isFunction(window.setRect)) {
        await window.setRect({ x: 0, y: 0, width: size.width, height: size.height });
    }
    else {
        await window.setPosition(0, 0);
        await window.setSize(size.width, size.height);
    }
}
exports.setWindowSize = setWindowSize;
async function getDriverInfo(driver) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    const session = await driver.getSession();
    const capabilities = await driver.getCapabilities();
    const desiredCapabilities = (_a = capabilities.get('desired')) !== null && _a !== void 0 ? _a : {};
    const platformName = (_c = (_b = capabilities.get('platformName')) !== null && _b !== void 0 ? _b : capabilities.get('platform')) !== null && _c !== void 0 ? _c : desiredCapabilities.platformName;
    const isMobile = ['android', 'ios'].includes(platformName === null || platformName === void 0 ? void 0 : platformName.toLowerCase());
    const info = {
        sessionId: session.getId(),
        isMobile,
        isNative: isMobile && !capabilities.get('browserName'),
        deviceName: (_d = desiredCapabilities.deviceName) !== null && _d !== void 0 ? _d : capabilities.get('deviceName'),
        platformName,
        platformVersion: capabilities.get('platformVersion'),
        browserName: (_e = capabilities.get('browserName')) !== null && _e !== void 0 ? _e : desiredCapabilities === null || desiredCapabilities === void 0 ? void 0 : desiredCapabilities.browserName,
        browserVersion: (_f = capabilities.get('browserVersion')) !== null && _f !== void 0 ? _f : capabilities.get('version'),
    };
    if (info.isNative) {
        const { Command } = require('selenium-webdriver/lib/command');
        let details;
        if (capabilities.has('viewportRect') && capabilities.has('pixelRatio') && capabilities.has('statBarHeight')) {
            details = {
                viewportRect: capabilities.get('viewportRect'),
                pixelRatio: capabilities.get('pixelRatio'),
                statBarHeight: capabilities.get('statBarHeight'),
            };
        }
        else {
            const getSessionDetailsCommand = new Command('getSessionDetails');
            details =
                process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3'
                    ? await driver.schedule(getSessionDetailsCommand)
                    : await driver.execute(getSessionDetailsCommand);
        }
        info.pixelRatio = details.pixelRatio;
        try {
            const getSystemBars = new Command('getSystemBars');
            const { statusBar, navigationBar } = process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3'
                ? await driver.schedule(getSystemBars)
                : await driver.execute(getSystemBars);
            info.statusBarHeight = statusBar.visible ? statusBar.height : 0;
            info.navigationBarHeight = navigationBar.visible ? navigationBar.height : 0;
        }
        catch (err) {
            info.statusBarHeight = (_j = (_g = details.statBarHeight) !== null && _g !== void 0 ? _g : (_h = details.viewportRect) === null || _h === void 0 ? void 0 : _h.top) !== null && _j !== void 0 ? _j : 0;
            info.navigationBarHeight = 0;
        }
    }
    return info;
}
exports.getDriverInfo = getDriverInfo;
async function getTitle(driver) {
    return driver.getTitle();
}
exports.getTitle = getTitle;
async function getUrl(driver) {
    return driver.getCurrentUrl();
}
exports.getUrl = getUrl;
async function visit(driver, url) {
    await driver.get(url);
}
exports.visit = visit;
async function takeScreenshot(driver) {
    return driver.takeScreenshot();
}
exports.takeScreenshot = takeScreenshot;
async function click(driver, element) {
    if (isSelector(element))
        element = await findElement(driver, element);
    await element.click();
}
exports.click = click;
async function hover(driver, element) {
    if (isSelector(element))
        element = await findElement(driver, element);
    if (process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3') {
        const { ActionSequence } = require('selenium-webdriver');
        const action = new ActionSequence(driver);
        await action.mouseMove(element).perform();
    }
    else {
        await driver.actions().move({ origin: element }).perform();
    }
}
exports.hover = hover;
async function type(driver, element, keys) {
    if (isSelector(element))
        element = await findElement(driver, element);
    await element.sendKeys(keys);
}
exports.type = type;
async function scrollIntoView(driver, element, align = false) {
    if (isSelector(element))
        element = await findElement(driver, element);
    await driver.executeScript('arguments[0].scrollIntoView(arguments[1])', element, align);
}
exports.scrollIntoView = scrollIntoView;
async function waitUntilDisplayed(driver, selector, timeout) {
    const { until } = require('selenium-webdriver');
    const element = await findElement(driver, selector);
    await driver.wait(until.elementIsVisible(element), timeout);
}
exports.waitUntilDisplayed = waitUntilDisplayed;
async function getOrientation(driver) {
    const { Command } = require('selenium-webdriver/lib/command');
    const getOrientationCommand = new Command('getOrientation');
    const orientation = process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3'
        ? await driver.schedule(getOrientationCommand)
        : await driver.execute(getOrientationCommand);
    return orientation.toLowerCase();
}
exports.getOrientation = getOrientation;
async function getElementRegion(_driver, element) {
    return element.getRect();
}
exports.getElementRegion = getElementRegion;
async function getElementAttribute(_driver, element, attr) {
    return element.getAttribute(attr);
}
exports.getElementAttribute = getElementAttribute;
async function getElementText(_driver, element) {
    return element.getText();
}
exports.getElementText = getElementText;
async function performAction(driver, steps) {
    const { Command } = require('selenium-webdriver/lib/command');
    const performTouchCommand = new Command('performTouch').setParameters({
        actions: steps.map((_a) => {
            var { action } = _a, options = __rest(_a, ["action"]);
            return ({ action, options });
        }),
    });
    if (process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3') {
        await driver.schedule(performTouchCommand);
    }
    else {
        await driver.execute(performTouchCommand);
    }
}
exports.performAction = performAction;
const browserOptionsNames = {
    chrome: 'goog:chromeOptions',
    firefox: 'moz:firefoxOptions',
};
async function build(env) {
    var _a;
    const { Builder } = require('selenium-webdriver');
    const parseEnv = require('@applitools/test-utils/src/parse-env');
    const { browser, capabilities, url, attach, proxy, configurable = true, appium = false, args = [], headless, } = parseEnv(Object.assign(Object.assign({}, env), { legacy: (_a = env.legacy) !== null && _a !== void 0 ? _a : process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION === '3' }));
    const desiredCapabilities = Object.assign({}, capabilities);
    if (configurable) {
        const browserOptionsName = browserOptionsNames[browser || desiredCapabilities.browserName];
        if (browserOptionsName) {
            const browserOptions = desiredCapabilities[browserOptionsName] || {};
            browserOptions.args = [...(browserOptions.args || []), ...args];
            if (headless)
                browserOptions.args.push('headless');
            if (attach) {
                browserOptions.debuggerAddress = attach === true ? 'localhost:9222' : attach;
                if (browser !== 'firefox')
                    browserOptions.w3c = false;
            }
            desiredCapabilities[browserOptionsName] = browserOptions;
        }
    }
    if (appium && browser === 'chrome') {
        desiredCapabilities['appium:chromeOptions'] = { w3c: false };
    }
    const builder = new Builder().withCapabilities(desiredCapabilities);
    if (url && !attach)
        builder.usingServer(url.href);
    if (proxy) {
        builder.setProxy({
            proxyType: 'manual',
            httpProxy: proxy.http || proxy.server,
            sslProxy: proxy.https || proxy.server,
            ftpProxy: proxy.ftp,
            noProxy: proxy.bypass,
        });
    }
    const driver = await builder.build();
    return [driver, () => driver.quit()];
}
exports.build = build;
exports.wrapDriver = legacy.wrapDriver;
